Create a new Upgrade.md document

pnax opened the pull request:
## Purpose

Separate installation and upgrade instructions in documentation.

## Context

This was discussed in #839.
Part of this PR addresses #618.

## Changes

* create a new `docs/Upgrade.md` document that acts as an entry point for the documentation on how to upgrade the Backend.

* add a title and a section to all specific upgrade documents

* move all upgrade documents into a subfolder `docs/upgrade`

* move all patch scripts into a subfolder `share/patch`

* remove unused modules from patch scripts


## How to test this PR

Install Zonemaster::Backend v6.2.0 or v7.0.0 then follow the Upgrade instructions to get to v8.0.0.

matsduf commented on docs/Installation.md:
```diff
@@ -784,8 +766,8 @@ Remove the database file and recreate it following the installation instructions
 [Post-installation]: #6-post-installation
-[README.md-upgrade]: /README.md#upgrade
 [Upgrade database]: #7-upgrade-zonemaster-database
+[upgrade document]: docs/Upgrade.md
```
It resides in the same directory so it should be (initial upper case is best):
```
[Upgrade document]: docs/Upgrade.md
```

pnax replied:
sould be fixed

matsduf commented on docs/Upgrade.md:
```diff
@@ -0,0 +1,45 @@
+
+To upgrade Zonemaster::Backend perform the following tasks:
+
+  1. stop the `zm-rpcapi` and `zm-testagent` daemons
```
In FreeBSD they are, for some reason, `zm_rpcapi` and `zm_testagent`

pnax replied:
Should we consider standardize how the daemons are called?

matsduf replied:
Yes, but is a breaking change how it is called. It is better with hyphen than underscore.

matsduf commented on docs/Upgrade.md:
```diff
@@ -0,0 +1,45 @@
+To upgrade Zonemaster::Backend perform the following tasks:
+
+  1. stop the `zm-rpcapi` and `zm-testagent` daemons
+  2. install the latest version from `cpanm` with `cpanm Zonemaster::Backend`
```
This will, I think, automatically upgrade Zonemaster::LDNS and Zonemaster::Engine too, which is good in one way, but if some installation parameter is used for that OS, then it is maybe less good.

pnax replied:
I understand the use case you are describing, however I don't see how this use use case was handled before.

Are you suggesting that we should consider adding a note for this specific use case? Or can we consider that if someone is in this use case that means that they have already moved apart from the installation instructions, so we can infer that they can continue to tweak the instructions?

matsduf replied:
Maybe a reference that Engine should be updated first, with a link to the Engine installation document (in the future maybe an Engine upgrade instruction).

Besides the possible parameters for Zonemaster::LDNS there is also the possibility that there are new dependencies for Zonemaster::LDNS or Zonemaster::Engine that should be installed first. (Zonemaster::LDNS is handled together with Zonemaster::Engine.)

matsduf commented on docs/Upgrade.md:
```diff
@@ -0,0 +1,45 @@
+ 1.0.3 ≤ version < 1.1.0            | [Upgrade to 1.1.0]   |
+ 1.1.0 ≤ version < 5.0.0            | [Upgrade to 5.0.0]   |
+ 5.0.0 ≤ version < 5.0.2            | [Upgrade to 5.0.2]   | For MySQL/MariaDB only
+ 5.0.2 ≤ version < 8.0.0            | [Upgrade to 8.0.0]   |
```
Do we go from v7.0.0 to v8.0.0? Or should it be v7.1.0? Do we have breaking changes after v7.0.0?

pnax replied:
The bump to v7.0.0 has been done in #826 which will land in master with next release. But latest fix release took the v7.0.0 version number. So I am just applying the same increase made in #826.

matsduf replied:
I am not sure if it is breaking, or just a feature change, i.e. v7.0.0->v7.1.0.

matsduf reviewed (commented):
These instructions are still limited to database changes and do not contain any other information.

I think it looks good. It is not more complex than before, and it clarifies to the users about upgrades.

matsduf reviewed (changes requested):
There are some changes needed.

pnax commented:
I've rebased on top of develop and applied some changes required. For the other ones, see their respective discussions.

pnax commented:
I added a link to the Engine installation instruction.

I also added a pointer to the installation instructions for each old upgrade document (by old I mean < v8.0.0). I agree that this is far from perfect (since the user will jump between installation documents) but this should cover must cases. If anyone come with a better idea, feel free to share it.

matsduf commented on docs/Installation.md:
```diff
@@ -468,7 +450,7 @@ Once the number reaches 100% a JSON object is printed and zmtest terminates.
 Zonemaster-Backend and keep the database, then you might have to upgrade the
 database to use it with the new version of Zonemaster-Backend. Please see the
-[upgrade][README.md-upgrade] information.
+[Upgrade document].
```
I think the section "Upgrade Zonemaster database" should be moved to the top just before the first installation section, and it should be renamed to "Upgrade of Zonemaster backend". Here it could be noted that you should remove the database first if you do not want to upgrade. Etc.

matsduf replied:
Any comment on this suggestion?

pnax commented on docs/Upgrade.md:
```diff
@@ -0,0 +1,51 @@
+
+To upgrade Zonemaster::Backend perform the following tasks:
+
+  1. stop the `zm-rpcapi` and `zm-testagent` daemons
```
this should land in #886 once this PR is merged.

matsduf replied:
If this PR does not assumes #886 it should be 
```
  1. stop the `zm-rpcapi` and `zm-testagent` daemons (`zm_rpcapi` and `zm_testagent` on FreeBSD)

matsduf commented on docs/Upgrade.md:
```diff
@@ -0,0 +1,57 @@
+
+  1. stop the `zm-rpcapi` and `zm-testagent` daemons
+  2. apply any instructions specific to this new release
+  3. install the latest version from `cpanm` with `cpanm Zonemaster::Backend`
```
```
  3. install the latest version from CPAN with `cpanm Zonemaster::Backend`
```

matsduf commented on docs/Upgrade.md:
```diff
@@ -0,0 +1,51 @@
+  1. stop the `zm-rpcapi` and `zm-testagent` daemons
+  2. install the latest version from `cpanm` with `cpanm Zonemaster::Backend`
+  3. apply any instructions specific to this new release
+  4. start the `zm-rpcapi` and `zm-testagent` daemons
```
```
  4. start the `zm-rpcapi` and `zm-testagent` daemons (`zm_rpcapi` and `zm_testagent` on FreeBSD)

matsduf reviewed (commented):
The upgrade instructions do not handle the following issues:

1. Updated default configuration (ini) file with new settings
2. Updated start script
3. If the start scripts  for FreeBSD are name changed by #886 then there should be instructions for removal of old scripts too

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
```diff
@@ -0,0 +1,78 @@
+# Upgrade to 8.0.0
+
```
````
On FreeBSD run the following command first to become root:

```sh
su -l
```

````

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
```diff
@@ -0,0 +1,78 @@
+
+## New dependencies
+
+Depending on the used OS, run the corresponding command.
```
Comment not needed.

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
```diff
@@ -0,0 +1,78 @@
+
+Depending on the used OS, run the corresponding command.
+
+### FreeBSD
```
I suggest that we preserve the order CentOS : Debian/Ubuntu : FreeBSD in all places.

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,78 @@
+Go to the share folder:
+```
+cd `perl -MFile::ShareDir=dist_dir -E 'say dist_dir("Zonemaster-Backend")'`
+```
````
I suggest that is moved into the command boxes below.

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,78 @@
+cd `perl -MFile::ShareDir=dist_dir -E 'say dist_dir("Zonemaster-Backend")'`
+```
+
+And then install the `zm-rpcapi` daemon (`zm_rpcapi` on FreeBSD).
````
The comment is not needed. There is a risk that you start doing that without seeing the commands below.

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,78 @@
+```
+su -l
+install -v -m 755 ./zm_rpcapi-bsd /usr/local/etc/rc.d/zm_rpcapi
+```
````
````
```sh
cd `perl -MFile::ShareDir=dist_dir -E 'say dist_dir("Zonemaster-Backend")'`
install -v -m 755 ./zm_rpcapi-bsd /usr/local/etc/rc.d/zm_rpcapi
```
````

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,78 @@
+
+```
+sudo install -v -m 755 ./zm-rpcapi.lsb /etc/init.d/zm-rpcapi
+```
````
````
```sh
cd `perl -MFile::ShareDir=dist_dir -E 'say dist_dir("Zonemaster-Backend")'`
sudo install -v -m 755 ./zm-rpcapi.lsb /etc/init.d/zm-rpcapi
```
````

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,78 @@
+```
+
+
+### MySQL (or MariaDB)
````
Rather "MariaDB (MySQL)" to be consistent.

matsduf commented on docs/Installation.md:
```diff
@@ -36,6 +35,11 @@
 This document contains all steps needed to install Zonemaster::Backend. For an overview of the
 Zonemaster product, please see the [main Zonemaster Repository].
 
+If you upgrade your Zonemaster installation with a newer version of
```
I still think there should be ha headline "Upgrade" above this text to make the reader catch that and not go to its OS directly.

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
```diff
@@ -0,0 +1,78 @@
+# Upgrade to 8.0.0
```
I think there should be information in this document saying that the default configuration file has changed.

pnax replied:
A new section has been added.

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
```diff
@@ -0,0 +1,78 @@
+## Upgrading the database
+
+If your Zonemaster database was created by a Zonemaster-Backend version smaller
+than v8.0.0, and not upgraded, use the following instructions.
```
Maybe add "Run the upgrade command with `sudo` if needed."

pnax replied:
Comment added.

blacksponge commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,78 @@
+```
+
+
+## Upgrading the database
````
Can you add that depending of the database size, this upgrade can take some time (zonemaster.net (1.4 million test results) will take around 30min)?

matsduf commented on docs/upgrade/upgrade_zonemaster_backend_ver_8.0.0.md:
````diff
@@ -0,0 +1,104 @@
+```sh
+su -l
+pkg install p5-Plack-Middleware-ReverseProxy
+```
````
This gives one block of two lines that you can paste at once. The first line is executed, and if correct root password is provided, then a new root shell is started. The second line is executed when the root shell has terminated, i.e. as the user before `su -l`. 

The following would give you what you want, but I suggest to have `su -l` as an introduction instead, especially if several root commands are to be executed.
````
```sh
su -l root -c 'pkg install p5-Plack-Middleware-ReverseProxy'
```
````

pnax replied:
Btw, can we envisage adding `sudo` on FreeBSD?
```
pkg install sudo
```
It would make it so much easier to maintain.

matsduf commented on docs/Installation.md:
```diff
@@ -25,25 +25,29 @@
 
+If you upgrade your Zonemaster installation with a newer version of
+Zonemaster::Backend instead, and want to keep the database, then see the
+[Upgrade document]. Otherwise [remove the database][Removing database] and
```
The link `[remove the database][Removing database]` does not work.

pnax commented:
Conflict fixed. Please re-review.
